### ISSUE TITLE
[lindholm] Inspection of Attributed Syntax Trees

### ISSUE THREAD

**Issue by grammarware** [grammarware/sle16#12]
Submitted by @gorelhedin to https://github.com/grammarware/sle16/tree/master/lindholm

**Comment by patrimrodilla** [comment]
This artifact allows visualizing attributed trees in AG’s inspection. I rigorously followed the steps documented in order to test the artifact. I think the first part, the environment and prerequisites configuration and installation of the ant (in case you don’t have it) and the Calc language demonstration is very well documented (step by step) and easy to follow. The second part, which includes the ExtendJ example is more complex but it’s also **well documented**.
I could suggest for future versions that, due to the specific environment of the artifact (only for mac, with jre/jdk and ant installation etc), which is very common, maybe it could be useful to develop an script with the environment configuration (environment variables including in the PATH etc), in order to facilitate these tasks. 

In terms of technological dependencies, I supposed that, as well as in the case of ExtendJ, it will need to prepare in some way any compiler in order to be support by the artifact. I understand this requirement, but maybe the authors could work in the future in a more high level process definition, in order to homogenize in some way these processes (it could be a future line of work). This could improve the applicability of the artifact to different compilers (and subsequently languages). The artifact showed is **consistent** in terms of the functionalities expressed in the paper: visualization as a graph of the AG attributed trees, filtering options and cached attributes.

I have some **suggestions** in terms of the graph visualization chosen: maybe could be interesting to employ all the information about each node that we have in order to improve the visualization: 
- it could be interesting to differ between types of nodes/relations (only declarations, or only constructs…) in a visualization way: by colors? by forms of the nodes?
- it could be interesting in order to use the artifact in teaching environment to present the original code in a more visible way near the graph, in terms of working on cognitive support to comparison tasks.
- it could be interesting to implement the filter systems in a none-declarative or programmatic way: for instance including buttons or other kind of interface elements for the most common filters applications, but maintaining the declarative filter for advance filters.

Regarding **utility**, the authors talks about teaching applications, learning AG and debugging issues as main uses. I agree with these considerations, although I have some concerns about the usability of the interface design employed (see my suggestions above). Also, it will useful to work in a more formal description of the possible users/stakeholders of these application: are experienced people with AG? or only programming students?

In summary, I think the artifact is well explained, developed and covers a need, so I would recommend its approval. Once said, it is necessary to work on InfoVis solutions (I also suggest to the authors to take a look to most modern InfoVis solutions for graphs) and also to perform some empirical user studies with user definition in order to evaluate the usability and complete usefulness of it.

**Comment by vnssousa** [comment]
This artifact does what it reports.

The documentation is clear to follow. The only issues to point are, when installing extendj my personal installation had one folder depth than the instructions leading to some minor mismatches with the instructions, but these do not prevent the user from following them. At some point in the guide the is the instruction to do a "cd extendj" followed by a "cd .." this does nothing and is probably an artifact of an earlier version of the guide. The guide should also report a dependency to Oracle Java specifically as I had trouble running it on my default installation that was OpenJDK.

Running the examples was straightforward and clear, with no issues to report. I would suggest adding fixed zooms to deal with larger diagrams. I know the point was to show the slowdown, and that with the filtering we regain performance, but zooming in at that size showed no change making it hard to differentiate between a failure or just normal behavior due to the size.

**Comment by gorelhedin** [comment]
@patrimrodilla  Thanks for the review and for your suggestions about visualization, interaction, and user studies. This is definitely an interesting area for future work. One thing we have been thinking about is indeed the possibility to create filters more interactively, rather than direct editing of the filter program.

@vnssousa  Thanks for the review and suggestions. We were not aware that there were problems in running the tool with OpenJDK, and we will look into this. We will also improve the guide so that the commands are easier to reuse. Thanks also for the suggestion for avoiding slowdown for large non-filtered ASTs.

**Comment by gorelhedin** [comment]
@vnssousa We have run the artefact now with OpenJDK, and it works with Java 8 update 40 or later. We will update the guide to mention this. What version of OpenJDK were you using? It would also be interesting to see the output from the failing run, in case you still have it around.

**Comment by vnssousa** [comment]
@gorelhedin I recreated the issue

~/SLE/lindholm/CalcExample $ java -jar ../DrAST.jar 
Error: Could not find or load main class drast.views.gui.DrASTGUI

~/SLE/lindholm/CalcExample $ java -version
openjdk version "1.8.0_91"
OpenJDK Runtime Environment (build 1.8.0_91-8u91-b14-3ubuntu1~16.04.1-b14)
OpenJDK 64-Bit Server VM (build 25.91-b14, mixed mode)

This error disappears when I use the hand installed Oracle JDK. But I guess it does have  to do with the version. This is the default-jdk version it gave-me with apt-get.

**Comment by grammarware** [comment]
On my side the artefact builds as expected:

```
Buildfile: /Users/zaytsev/projects/sle16/lindholm/CalcExample/build.xml

scanner-gen:
    [mkdir] Created dir: /Users/zaytsev/projects/sle16/lindholm/CalcExample/src/gen
    [jflex] Generated: LangScanner.java

parser-gen:
   [beaver] parser.beaver: Generated: LangParser.java

ast-gen:
  [jastadd] generating node types and weaving aspects
  [jastadd] done

gen:

build:
    [mkdir] Created dir: /Users/zaytsev/projects/sle16/lindholm/CalcExample/ant-bin
    [javac] Compiling 26 source files to /Users/zaytsev/projects/sle16/lindholm/CalcExample/ant-bin
    [javac] warning: [options] bootstrap class path not set in conjunction with -source 1.7
    [javac] Note: Some input files use unchecked or unsafe operations.
    [javac] Note: Recompile with -Xlint:unchecked for details.
    [javac] 1 warning

test:
    [junit] Testsuite: tests.ParseTests
    [junit] Tests run: 7, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.031 sec
    [junit] 
    [junit] Testsuite: tests.TestDumpTree
    [junit] Tests run: 14, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.102 sec
    [junit] 
    [junit] Testsuite: tests.TestNameAnalysis
    [junit] Tests run: 10, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.058 sec
    [junit] 
    [junit] Testsuite: tests.TestPrettyPrinting
    [junit] Tests run: 12, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.057 sec
    [junit] 

BUILD SUCCESSFUL
Total time: 3 seconds
```

And the resulting jar does not give the error reported above — perhaps I got lucky by testing it on a relatively fresh machine without many JDKs preinstalled and getting in one another's way. The main part of the evaluation came after the build, when I could run the visualiser and work with it interactively, changing the filters, defining your own ones, etc. The base paper of this artefact is a tool paper, and it basically serves as an extended and detailed description of the tool which is the artefact itself. By following the paper literally I was able to reproduce everything reported there with extreme ease, and then playing with filters, visualisation itself (zooming, moving elements, writing my own ExtendJ-based code) also went smoothly, modulo my general understanding of JastAdd yet the lack of expertise in it. Accepting this artefact and attaching it to the paper in the ACM DL, would help future readers not just get the text describing how it at some point worked, but also run the tool and live through all the explanations, which is a big plus and a good step towards ideal reality.

**Comment by grammarware** [comment]
The Artefact Evaluation Committee of SLE 2016 has reached the conclusion yesterday during the committee meeting that this artefact conforms to the expectations and is **approved** for inclusion as auxiliary material for the publication.

Feel free to make some appropriate adjustments before submitting the materials to ACM, @gorelhedin, and enjoy your seal of approval.

**Comment by gorelhedin** [comment]
Thanks for the approval and all comments. I have updated the artefact and uploaded as a zip on ACM.
